lib/private/Preview/Generator.php

```diff
@@ -426,10 +426,16 @@ private function getCachedPreview(ISimpleFolder $previewFolder, $width, $height,
 	 * @return ISimpleFolder
 	 */
 	private function getPreviewFolder(File $file) {
+		$folderName = implode('/', str_split($file->getId()));
```

Comment:
I like the idea :+1: 

Not sure about the file id as value. The length is not fixed and most of the files are in 1/ then. Think about 1000+ and 10000+ as file ids. There is a column `path_hash` which looks promising but there is no `getPathHash` method.

Reply:
So Im not against the idea but we'd need some migraion as well. As otherwise you end up with unusable previews.

Also i would indeed not use the fileid but a hash or something. Not the path hash (as it is not stable). But something that does uniformly distribute the previews...

Reply:
@kesselb it may sound crazy, but could we sha1 the file ID and then consider the first 7 digits. It's how git identifies the commits, right? it's just an idea, open to other hashing algorithms (sha1 isn't doing too well in these days).

@rullzer is a migration tool really required? that's why I call getFolder twice, the first one with the new hierarchy and the second one with the old one. I thought that was enough to not break the old previews access.

Reply:
Sure. We should pick a fast algorithm. Mind to do some tests with md5 + str_split(id, 8)? This should result in 4 folders each with 8 characters. But it's not a uniformly distributed value.

Reply:
What I meant actually is to hash the file ID, then use the first X digits and build a tree of folders in which each foldername is composed of one the hash hexdigit.

Something like hash(fileID) = 5005943ca608c93225e684dbfb825973ba02dbf9 => we consider only 5005943:

5 -> 0 -> 0 -> 5 -> 9 -> 4 -> 3 -> previews

I guess this should provide an uniform distribution.

Reply:
We are talking about the same concept. Only the details are different ;)

```suggestion
		$folderName = implode('/', str_split(substr(md5($file->getId()), 0, 7)));
```

Reply:
@kesselb so, I tried replacing the file ID with the substr of sha1 and everything worked fine, with a fairly good distribution.
Also, retrocompatibility is working thanks to the double try-catch.

What do you think?

Reply:
Fine by me. Actually we neither need `md5` or `sha1`. `crc32` should be enough but according to https://stackoverflow.com/questions/16122067/md5-vs-crc32-which-ones-better-for-common-use it's slower than md5.  

@rullzer md5, sha1 or something else?

Code looks a bit cluttered right now. We can improve this by using return.
```

		$folderName = implode('/', str_split($file->getId()));

		try {
			return $this->appData->getFolder($folderName);
		} catch (NotFoundException $e) {
			// preview folder does not exist.
		}

		try {
			// In order to have some retrocompatibility
			return $this->appData->getFolder($file->getId());
		} catch (NotFoundException $e) {
			// preview folder does not exist.
		}

		return $this->appData->newFolder($folderName);
```

Reply:
Or just inverse the file id. If it is ascending integer, it should be more or less equally distributed.
`10422` -> `2/2/4/...`

Reply:
@tflidd true, but then we would have problems with ids with different length,, wouldn't we?

Reply:
@kesselb @rullzer any update on this?

Reply:
> I tried replacing the file ID with the substr of sha1 and everything worked fine, with a fairly good distribution.

Are you going to push this? I still think we should go with md5 (because cheaper) but sha1 also works.

Reply:
@kesselb We can totally switch to md5... I was just waiting for a final decision on this

Reply:
I find https://stackoverflow.com/questions/996843/when-is-crc-more-appropriate-to-use-than-md5-sha1 very interesting. Please pick some method and push the changes. We can still change it later. The next step would be to look what the CI thinks about this.